File: VisualBasicDotNet/99beers.vb
Imports System

Module Program
    Dim bottles As Integer = 99

    Function randomNum()
        Dim rand As New Random()
        return rand.Next(0, 101)
    End Function

    Sub Main(args As String())
        While Not bottles = 0
            Dim ifWaste As Boolean = false
            While 1 < bottles Or bottles <= -1
                verse(bottles)
                bottles = bottles - 1
            End While

            Console.WriteLine($"{bottles} bottle of beer on the wall,")
            Console.WriteLine($"{bottles} bottle of beer.")
            If randomNum() > 80 Then
                Console.WriteLine("If that one bottle should happens to fall,")
                If randomNum() < 20 Then
                ifWaste = true
                End if
            Else
                Console.WriteLine("Take one down and pass it around.")
            End If

            bottles = ending(bottles, ifWaste)
        End While
    End Sub

    Sub verse(bottles As Integer)
        Console.WriteLine($"{bottles} bottle of beer on the wall,")
        Console.WriteLine($"{bottles} bottle of beer.")
        If randomNum() > 80 Then
            Console.WriteLine("If one of those bottles should happen to fall,")
            Console.WriteLine($"{bottles - 1} bottles of beer on the wall.")
            Console.WriteLine()
        Else If randomNum() < 20 Then
            Console.WriteLine("If that one bottle should happens to fall,")
            Console.WriteLine("what a waste of alcohol!")
            Console.WriteLine()
        Else
            Console.WriteLine("Take one down and pass it around.")
            Console.WriteLine($"{bottles - 1} bottles of beer on the wall.")
            Console.WriteLine()
        End If
    End Sub

    Function infiniteBottles()
        verse(0)
        return -1
    End Function

    Function ending(bottles As Integer, waste As Boolean)
        If randomNum() = 100 And bottles = 1 Then
            If waste Then
                Console.WriteLine("what a waste of alcohol!")
                Console.WriteLine()
            Else
                Console.WriteLine("No more bottles of beer on the wall.")
                Console.WriteLine()
            End If

            bottles = infiniteBottles()
        Else
            If waste Then
                Console.WriteLine("what a waste of alcohol!")
                Console.WriteLine()
            Else
                Console.WriteLine("No more bottles of beer on the wall.")
                Console.WriteLine()
            End If

            Console.WriteLine("No more bottles of beer on the wall,")
            Console.WriteLine("no more bottles of beer.")
            If randomNum() <= 60 Then
                bottles = 0
                If randomNum() <= 10 Then
                    Console.WriteLine("There's nothing else to fall,")
                    Console.Write("because there's no more bottles of beer on ")
                    Console.WriteLine("the wall.")
                    Console.WriteLine()
                Else
                    Console.Write("We've taken them down and passed them")
                    Console.WriteLine("around;")
                    Console.WriteLine("now we're drunk and passed out!")
                    Console.WriteLine()
                End If
            Else
                bottles = 99
                Console.WriteLine("Go to the store and buy some more,")
                Console.WriteLine($"{bottles} bottles of beer on the wall.")
                Console.WriteLine()
            End If
        End If
        
        return bottles
    End Function
End Module

File: CPlusPlus/99beers.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>

int bottles = 99;

int random()
{
  srand(time(0));
  return (rand() % 100) + 1;
}

void verse(int bottles)
{
  std::cout << bottles << " bottles of beer on the wall,\n";
  std::cout << bottles << " bottles of beer.\n";
  if (random() > 80)
  {
    std::cout << "If one of those bottles should happen to fall,\n";
    std::cout << bottles - 1 << " bottles of beer on the wall.\n\n";
  }
  else if (random() < 20)
  {
    std::cout << "If that one bottle should happens to fall,\n";
    std::cout << "what a waste of alcohol!\n\n";
  }
  else
  {
    std::cout << "Take one down and pass it around.\n";
    std::cout << bottles - 1 << " bottles of beer on the wall.\n\n";
  }
}

int infiniteBottles()
{
  verse(0);
  return -1;
}

int ending(int bottles, bool waste)
{
  if ((random() == 100) && (bottles == 1))
  {
    if (waste)
    {
      std::cout << "what a waste of alcohol!\n\n";
    }
    else
    {
      std::cout << "No more bottles of beer on the wall.\n\n";
    }
    bottles = infiniteBottles();
  }
  else
  {
    if (waste)
    {
      std::cout << "what a waste of alcohol!\n\n";
    }
    else
    {
      std::cout << "No more bottles of beer on the wall.\n\n";
    }

    std::cout << "No more bottles of beer on the wall,\n";
    std::cout << "no more bottles of beer.\n";
    if (random() <= 60)
    {
      bottles = 0;
      if (random() <= 10)
      {
        std::cout << "There's nothing else to fall,\n";
        std::cout << "because there's no more bottles of beer on the wall."
                  << "\n\n";
      }
      else
      {
        std::cout << "We've taken them down and passed them around;\n";
        std::cout << "now we're drunk and passed out!\n\n";
      }
    }
    else
    {
      bottles = 99;
      std::cout << "Go the the store and buy some more,\n";
      std::cout << bottles << " bottles of beer on the wall.\n\n";
    }
  }

  return bottles;
}

int main()
{
  while (bottles != 0)
  {
    bool ifWaste = false;
    while (1 < bottles || bottles <= -1)
    {
      verse(bottles);
      bottles--;
    }

    std::cout << bottles << " bottle of beer on the wall,\n";
    std::cout << bottles << " bottle of beer.\n";
    if (random() > 80)
    {
      std::cout << "If that one bottle should happens to fall,\n";
      if (random() < 20)
      {
        ifWaste = true;
      }
    }
    else
    {
      std::cout << "Take one down and pass it around.\n";
    }

    bottles = ending(bottles, ifWaste);
  }
}
